Feature/oauth self hosted

acoumb opened the pull request:
Current PR contains the setup required to setup OAuth flow through a self hosted engine for following integrations: Dynamics, Semrush, Google Search Console, Hubspot.

The setting `UseUmbracoAuthorization` will toggle the OAuth flow from the default Umbraco one to a private one.

The custom settings go under the `[Integration]OAuthSettings` and contain the following properties:
- client id
- client secret
- redirect uri
- scopes
- authorization endpoint
- token endpoint
Additional keys have been added to support integrations running on V8.

For authorization flow I am using two services, the Umbraco default one and one custom. Both implement `I[Integration]AuthorizationService` with methods to retrieve the authorization URL, the access token
and refresh token.

A factory method will be then used to load the proper authorization service.

AndyButland commented on src/Umbraco.Cms.Integrations.SEO.GoogleSearchConsole.UrlInspectionTool/Configuration/GoogleSearchConsoleSettings.cs:
```diff
@@ -0,0 +1,25 @@
+
+        public string InspectUrl { get; set; }
+
+        public bool UseUmbracoAuthorization { get; set; }
```
```suggestion
        public bool UseUmbracoAuthorization { get; set; } = true;
```

AndyButland commented on src/Umbraco.Cms.Integrations.SEO.Semrush/Configuration/SemrushSettings.cs:
```diff
@@ -1,29 +1,23 @@
-        public static string TokenDbKey = "Umbraco.Cms.Integrations.Semrush.TokenDbKey";
-
-        public static string AllowLimitOffsetHeaderName = "Allow-Limit-Offset";
+        public bool UseUmbracoAuthorization { get; set; }
```
```suggestion
        public bool UseUmbracoAuthorization { get; set; } = true;
```

AndyButland commented on src/Umbraco.Cms.Integrations.Crm.Dynamics/Configuration/DynamicsSettings.cs:
```diff
@@ -8,13 +8,19 @@ public class DynamicsSettings
 
         public string ApiPath { get; set; }
+
+        public bool UseUmbracoAuthorization { get; set; }
```
```suggestion
        public bool UseUmbracoAuthorization { get; set; } = true;
```

AndyButland reviewed (approved):
Looks great, I think this will be a nice option for users to have.  Other than a few minor inline comments, the only thing missing that I can see before you merge is to add some clear documentation of the two options on the readme pages for each integration.

Maybe you could do a quick demo in the refining session we've got later too?

acoumb commented on src/Umbraco.Cms.Integrations.Crm.Dynamics/DynamicsComposer.cs:
```diff
@@ -17,15 +17,28 @@ namespace Umbraco.Cms.Integrations.Crm.Dynamics
             builder.AddNotificationHandler<UmbracoApplicationStartingNotification, UmbracoAppStartingHandler>();
 
-            builder.Services.AddSingleton<IAuthorizationService, AuthorizationService>();
+            builder.Services.AddTransient<UmbracoAuthorizationService>();
```
Indeed it doesn't make sense to serve a new instance every time, works fine as `Singleton`.

acoumb commented:
Feedback addressed, I am leaving this open for now so you can check also on my latest comment on the OAuth Proxy update [here](https://github.com/umbraco/Umbraco.Forms.Integrations/pull/70#issuecomment-1483756848).

acoumb commented:
- Included the updates for _Shopify_. 
- Removed the `AuthorizationEndpoint` property, as the service's authorization endpoint is added to the `BaseAuthorizationService`.
- Corrected the authorization controllers for Umbraco 8 to use `System.Web.Http` for the API calls on redirects.
